gtsfm/frontend/retriever/vocab_retriever.py

```diff
@@ -0,0 +1,84 @@
+"""Vocabulary Based image retriever for front-end.
+Authors: Jon Womack
+"""
+from collections import defaultdict
+from typing import List, Tuple
+
+import numpy as np
+import cbir
+import dask
+from dask.delayed import Delayed
+import networkx as nx
+import pickle
+from pathlib import Path
+
+from gtsfm.frontend.retriever.retriever_base import RetrieverBase
+import gtsfm.utils.logger as logger_utils
+
+
+logger = logger_utils.get_logger()
+
+
+class VocabTreeRetriever(RetrieverBase):
+    """Class for vocabulary based image retrieval. Selects a subset of the image pairs from Dataloader."""
+
+    def __init__(self, image_pair_indices):
+        """Initialize the Retriever.
+        Args:
+            image_pair_indices: All possible image pairs (image_id1, image_id2) as determined by Dataloader.
```

Comment:
another option would be to provide `num_images` as the constructor arg, instead. If we have the Retriever determine the image pairs, then the Loader shouldn't do so, IMHO.